File: Python/Valid Parentheses.py
# Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

# The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.


class Solution:
    # @return a boolean
	def isValid(self, s):
		if len(s) == 0:
			return False
		stack = []
		stack.append(s[0])
		for i in range(1,len(s)):
			if s[i] == "(" or s[i] == "[" or s[i] == "{":
				stack.append(s[i])
				continue
			else:
				if len(stack) == 0:
					return False
				current = stack.pop()
				if s[i] == ")" and current != "(":
					return False
				if s[i] == "]" and current != "[":
					return False
				if s[i] == "}" and current != "{":
					return False
		if len(stack) != 0:
			return False
		return True

File: Java/Valid Parentheses.java
class Valid Parentheses{
	public boolean isValid(String s) {
        int index = 0;
        LinkedList<Character> stack = new LinkedList<Character>();
        while(index<s.length()){
        	if(s.charAt(index) == '(' || s.charAt(index) == '{' || s.charAt(index) == '['){
        		stack.push(s.charAt(index));
        		index++;
        	}
        	else if(s.charAt(index) == ')'){
        		if(stack.isEmpty() || stack.peek() != '(')
        			return false;
        		else{
        			stack.pop();
        			index++;
        		}
        	}
        	else if(s.charAt(index) == ']'){
        		if(stack.isEmpty() || stack.peek() != '[')
        			return false;
        		else{
        			stack.pop();
        			index++;
        		}

        	}
        	else if(s.charAt(index) == '}'){
        		if(stack.isEmpty() || stack.peek() != '{')
        			return false;
        		else{
        			stack.pop();
        			index++;
        		}
        	}
        }
        if(stack.isEmpty())
            return true;
        else
            return false;
    }
}